### ISSUE TITLE
Improved Logging Interface

### ISSUE THREAD

**Issue by dansiegel** [PrismLibrary/Prism#588]
The existing logging interface ILoggerFacade provides very limited use with it's single Log method. Since this requires a string to be passed in for the message, this limits developers from implementing for more powerful loggers as the only real customization you can make is deciding whether to ignore the message based on it's priority/type, or whether to save it to a file, or send a message to a web service. I understand we can't exactly just add a new overloaded signature to the ILoggerFacade as this would be a breaking change for a lot of existing apps, so I would propose we add a new logging interface across Prism and deprecate the existing ILoggerFacade.

I'm thinking the new logger should look something like:

```cs
public interface IPrismLogger
{
    void Log( string message, Category category, Priority priority );
    void Log( object obj, Category category, Priority priority );
    void Log( string message, object obj, Category category, Priority priority );
    void Log( Exception ex, Priority priority );
}
```

This would provide a more robust logging interface that Prism could take advantage of across the different platforms. By simply passing in something like an Exception, this would allow a base implementation to have an output like:

```
MyException: Some Message
```

While still allowing a developer to decide that they want to provide more information about the exception like the StackTrace, and dig into any InnerExceptions.

**Comment by brianlagunas** [comment]
This is going to be a hard sell.  Given a developer can provide their own extension methods to extend what they pass to their logger, there must really be a motivating benefit to changing the current logger interface.  Besides creating this new interface, we would have to deprecate the old one, then eventually remove it which would be a breaking change for a lot of apps out there.

I will let the community weigh in on this.

**Comment by briannoyes** [comment]
The ILoggerFacade was intentionally simple with the intent that people using Prism would at least want to be able to collect some logging of what the framework itself was doing, but we didn't want to force any unnecessary complexity on people using it if their logging needs were simple. There is nothing stopping you from having whatever separate logging abstraction you want as a separate logging service, and only use the ILoggerFacade implementation to keep the simple string based output of the Prism libraries going to your logging store as well. 

I am in agreement with Brian L - I think the potential for breakage, adding complexity, trying to agree on what the extra parameters should be and so on are not worth it. Log whatever you want from your own code using your own logging service. Then just provide a simple implementation of ILoggerFacade that sends it stuff into your logging store as well. But most of that is initialization debug info anyway, unlikely to be important in a production environment.

**Comment by bkaankose** [comment]
I agree on all the complexity thing, but just a single string logging method is kind of too simple for logging purposes imo. Log method should have at least an override with Exception type and even maybe overrides with Category and Priority selected by default.

I don't say those won't increase complexity, but at the level of can be ignored.

**Comment by briannoyes** [comment]
I'm agreeing that a single string is too simple - for the rest of your app code. But there is no reason the ILoggerFacade has to constrain you on anything other than what Prism already logs in the initialization. And we are not going to modify all of the places that the Prism code does logging - which is not very many places to begin with - to fit into any single logging framework's patterns because that would constrain anyone who uses Prism to follow those patterns. 

That logging that Framework does is not much more than initialization code. Useful when you are getting your app up and running the first time, and also useful if you are dynamically plugging in modules at runtime if something fails. But it has zero bearing on what you choose to log and how you choose to log it from your own application code... that is completely up to you and whatever ILogXYZ you choose to inject into your own code to do that logging using whatever logging patterns you choose for your code. 

So I'm kind of missing the point on why the Prism ILoggerFacade needs to change?

**Comment by dansiegel** [comment]
@briannoyes this steams from issues #576 & #587. I totally agree that if the point is to provide more robust logging features within my app, then I could just add some extension methods to suit my needs. The problem really comes down to helping developers debug exceptions. As @brianlagunas and I discussed offline, you have some instant problems with logging an exception. First of all, different applications have different requirements. Second, different developers have different ideas about what needs to be logged. 

We could of course just catch the exceptions in Prism and do something like log "MyException - My Error Message" and send that as a string to the existing Log method. The problem with that though is that, is that you prevent developers from capturing that exception within the Prism modules themselves, to log them in the way that the developer requires. Speaking purely from my own experience on iOS in particular, the platform has a way of sometimes swallowing up exceptions. Without the right toolset to log issues like exceptions, it can make it very challenging to discover what's really causing the problem.

I don't think that all of the places where the logger is being used currently will necessarily change, which is why I included a Log method with the same signature to the ILoggerFacade in my proposal above. For me the issue is really the need to be logging exceptions thrown inside the Prism modules, since ILoggerFacade cannot be altered in any way as it would be breaking for too many WPF apps, and adding an extension within Prism would really prevent a developer from overriding the behavior the only worthwhile option is to introduce a new logging interface to replace ILoggerFacade.

**Comment by brianlagunas** [comment]
Given that the loggers primary function is to simply log Prism events, and the majority of apps that require more verbose logging wind up using a different logger and extension methods can be used to provide a parameter override to format things like exceptions as the dev needs, this is not something we will be doing anytime soon.
```
    public static class LoggerExtensions
    {
        public static void Log(this ILoggerFacade logger, Exception ex)
        {
            logger.Log(...);
        }
    }
```

**Comment by pspattillo** [comment]
“I am in agreement with Brian L - I think the potential for breakage, adding complexity, trying to agree on what the extra parameters should be and so on are not worth it. Log whatever you want from your own code using your own logging service.”Then just provide a simple implementation of ILoggerFacade that sends it stuff into your logging store as well. But most of that is initialization debug info anyway, unlikely to be important in a production environment.”

There are approaches for adding to it. I’ve appreciated integration of product logging and product consumer logging together. Integrated logs with control and filtering helps to link cause and effect and interplay between caller, called and resulting problems. That would be easy enough with a base interface cast to ILoggerFacade or ILogger but I was wondering if there’s a way to replace the implementation or conform it to one’s needs all the better. I don’t recall an architected way to replace it.

Is there an architected callback among those in PrismApplication or base classes? No real work for your team in that. Hope for the best and plan for the worst, you know?

Over the years I’ve accumulated a suite of tools that may be needed in the rarest of occasion but have saved my ass every so often—things like object serialization formatting, overloads that will interject a stack trace answering questions about problem was set up and brute force logging via interception as a last resort.

Are there any examples of replacing the logging implementation via the container? I’d like the ability to integrate output that can be viewed in context with PRISM 7+. I realize this is old but hopefully someone will see it.

Basically I’m just lazy and have been faced with correlating multiple logs in the past and it was not only a pain but counter-productive.
